lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -414,19 +343,243 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+    // Note: If two threads in the same warp write to the same shmem offset, the
+    // hardware resolves that without a stall or a bank conflict.  Therefore we
+    // don't need to avoid duplicate writes.
+    LinearLayout shmemStoreLayout = srcLayout.invertAndCompose(sharedLayout);
+    const int shmemAllocatedNumElems =
+        getTotalSize<unsigned>(scratchConfig.paddedRepShape);
+    assert(shmemStoreLayout.getOutDimSize(kOffset) <= shmemAllocatedNumElems);
+
+    // Layout for the load from shmem to registers.
+    LinearLayout shmemLoadLayout = dstLayout.invertAndCompose(sharedLayout);
+
+    // Check that the `register` fully determines the `iteration`.  That is,
+    // each thread does exactly the same reads and writes to shmem on each
+    // iteration, just with different input/output registers.
+    assert(
+        shmemStoreLayout.sublayoutIsZero({kLane, kWarp, kBlock}, {kIteration}));
+    assert(
+        shmemLoadLayout.sublayoutIsZero({kLane, kWarp, kBlock}, {kIteration}));
+
+    // iteration -> registers
+    SmallVector<SmallVector<int>> inRegsForIter =
+        collectRegsForIter(ctx, shmemStoreLayout);
+    SmallVector<SmallVector<int>> outRegsForIter =
+        collectRegsForIter(ctx, shmemLoadLayout);
+
+    // Avoid emitting a call to getClusterCTAId if we know it's always 0.
+    std::optional<Value> ctaId;
+    if (conversion.getInDimSize(kBlock) > 1)
+      ctaId = targetInfo.getClusterCTAId(rewriter, loc);
+
+    // Map src/dst registers to each iteration.
+    Value smemBase =
+        LLVM::getSharedMemoryBase(loc, rewriter, op.getOperation());
+    Type elemTy = inVals[0].getType();
+    auto outSize = shmemLoadLayout.getInDimSize(kRegister);
+    auto iterations = sharedLayout.getInDimSize(kIteration);
+
+    auto getVecAddrAndBlock = [&](ArrayRef<Value> offsetAndBlock)
+        -> std::pair<Value, std::optional<Value>> {
+      std::optional<Value> block =
+          isCrossCTA ? std::make_optional(offsetAndBlock[2]) : std::nullopt;
+      Value offset = offsetAndBlock[0];
+      if (scratchConfig.paddedSize > 0) {
+        assert(llvm::isPowerOf2_32(scratchConfig.paddedStride));
+        assert(llvm::isPowerOf2_32(scratchConfig.paddedSize));
+        auto rshiftVal = llvm::Log2_32(scratchConfig.paddedStride);
+        auto lshiftVal = llvm::Log2_32(scratchConfig.paddedSize);
+        offset = add(shl(lshr(offset, i32_val(rshiftVal)), i32_val(lshiftVal)),
+                     offset);
+      }
+      auto vecAddr = gep(sharedPtrTy, elemTy, smemBase, offset);
+      vecAddr.setInbounds(true);
+      return {vecAddr, block};
+    };
+
+    // register idx -> Value
+    llvm::MapVector<int, Value> outVals;
+    for (int i = 0; i < iterations; i++) {
+      if (i != 0)
+        barrier();
+
+      auto &inRegs = inRegsForIter[i];
+      auto &outRegs = outRegsForIter[i];
+
+      // XXX: CSE applyLinearLayout.
+      for (int j = 0; j < inVals.size() / iterations;
+           j += scratchConfig.inVec) {
```

Comment:
do we ever assert that scratchConfig.inVec * iterations == inVals.size()?  Same for outSize.

Reply:
I can do that

Reply:
Actually it should be `scratchConfig.inVec * iterations <= inVals.size()` because each iteration here handles a `repShape`, which has handles multiple `inVec`s.